File: tools/geo.mjs
/**
 * JS twin of GhostRaceKit's Geo.swift and GhostEngine.interpolatedDistance.
 * The Swift code cannot be compiled in every environment GhostRace is
 * developed in, so this module implements the identical algorithms and the
 * shared `crosscheck.json` fixture pins both implementations to the same
 * expected values. If you change one, change the other.
 */

export const EARTH_RADIUS_M = 6_371_000;

const rad = (deg) => (deg * Math.PI) / 180;

/** Great-circle (haversine) distance in meters. */
export function distanceM(a, b) {
  const dLat = rad(b.lat - a.lat);
  const dLon = rad(b.lon - a.lon);
  const lat1 = rad(a.lat);
  const lat2 = rad(b.lat);
  const h =
    Math.sin(dLat / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLon / 2) ** 2;
  return 2 * EARTH_RADIUS_M * Math.asin(Math.min(1, Math.sqrt(h)));
}

/** Cumulative meters at each polyline vertex; first 0, last = total length. */
export function cumulativeDistances(polyline) {
  const out = [];
  let total = 0;
  for (let i = 0; i < polyline.length; i++) {
    if (i > 0) total += distanceM(polyline[i - 1], polyline[i]);
    out.push(total);
  }
  return out;
}

/**
 * Project a point onto a polyline using a per-edge local equirectangular
 * projection. Returns { distanceAlongM, offsetM }.
 */
export function project(point, polyline, cumulative = cumulativeDistances(polyline)) {
  if (polyline.length < 2) throw new Error("polyline needs at least 2 points");
  let best = { distanceAlongM: 0, offsetM: Number.POSITIVE_INFINITY };
  for (let i = 0; i < polyline.length - 1; i++) {
    const a = polyline[i];
    const b = polyline[i + 1];
    const meanLat = rad((a.lat + b.lat) / 2);
    const mPerDegLat = (Math.PI / 180) * EARTH_RADIUS_M;
    const mPerDegLon = mPerDegLat * Math.cos(meanLat);

    const bx = (b.lon - a.lon) * mPerDegLon;
    const by = (b.lat - a.lat) * mPerDegLat;
    const px = (point.lon - a.lon) * mPerDegLon;
    const py = (point.lat - a.lat) * mPerDegLat;

    const lenSq = bx * bx + by * by;
    const t = lenSq === 0 ? 0 : Math.max(0, Math.min(1, (px * bx + py * by) / lenSq));
    const cx = t * bx;
    const cy = t * by;
    const offset = Math.hypot(px - cx, py - cy);
    if (offset < best.offsetM) {
      const edgeLen = cumulative[i + 1] - cumulative[i];
      best = { distanceAlongM: cumulative[i] + t * edgeLen, offsetM: offset };
    }
  }
  return best;
}

/** Coordinate at `d` meters along the polyline (inverse of projection). */
export function positionAtDistance(polyline, cumulative, d) {
  if (d <= 0) return { ...polyline[0] };
  const total = cumulative[cumulative.length - 1];
  if (d >= total) return { ...polyline[polyline.length - 1] };
  let i = 0;
  while (cumulative[i + 1] < d) i++;
  const f = (d - cumulative[i]) / (cumulative[i + 1] - cumulative[i]);
  const a = polyline[i];
  const b = polyline[i + 1];
  return { lat: a.lat + f * (b.lat - a.lat), lon: a.lon + f * (b.lon - a.lon) };
}

/**
 * Linear interpolation of distance-along-course at elapsed time t, clamped to
 * the first/last points. Twin of GhostEngine.interpolatedDistance.
 */
export function interpolatedDistance(t, points) {
  if (points.length === 0) return 0;
  const first = points[0];
  if (t <= first.t) return first.d;
  const last = points[points.length - 1];
  if (t >= last.t) return last.d;
  let lo = 0;
  let hi = points.length - 1;
  while (hi - lo > 1) {
    const mid = (lo + hi) >> 1;
    if (points[mid].t <= t) lo = mid;
    else hi = mid;
  }
  const a = points[lo];
  const b = points[hi];
  if (b.t <= a.t) return a.d;
  return a.d + ((t - a.t) / (b.t - a.t)) * (b.d - a.d);
}

File: ios/GhostRaceKit/Sources/GhostRaceKit/Geo.swift
import Foundation

/// A latitude/longitude pair. GhostRaceKit deliberately avoids CoreLocation so
/// the domain logic stays pure, portable, and testable anywhere.
public struct Coordinate: Codable, Hashable, Sendable {
    public var lat: Double
    public var lon: Double

    public init(lat: Double, lon: Double) {
        self.lat = lat
        self.lon = lon
    }
}

/// Result of projecting a GPS fix onto a segment's polyline.
public struct PolylineProjection: Equatable, Sendable {
    /// Meters along the polyline from its start to the closest point.
    public var distanceAlongM: Double
    /// Perpendicular distance from the fix to the polyline, in meters.
    /// Large values mean the athlete is off-course (or GPS is drifting).
    public var offsetM: Double
}

public enum Geo {
    public static let earthRadiusM = 6_371_000.0

    /// Great-circle distance between two coordinates, in meters.
    public static func distanceM(_ a: Coordinate, _ b: Coordinate) -> Double {
        let dLat = (b.lat - a.lat) * .pi / 180
        let dLon = (b.lon - a.lon) * .pi / 180
        let lat1 = a.lat * .pi / 180
        let lat2 = b.lat * .pi / 180
        let h = sin(dLat / 2) * sin(dLat / 2) + cos(lat1) * cos(lat2) * sin(dLon / 2) * sin(dLon / 2)
        return 2 * earthRadiusM * asin(min(1, sqrt(h)))
    }

    /// Cumulative distance in meters at each vertex of `polyline`.
    /// First entry is always 0; last entry is the polyline's total length.
    public static func cumulativeDistances(_ polyline: [Coordinate]) -> [Double] {
        var out: [Double] = []
        out.reserveCapacity(polyline.count)
        var total = 0.0
        for (i, point) in polyline.enumerated() {
            if i > 0 { total += distanceM(polyline[i - 1], point) }
            out.append(total)
        }
        return out
    }

    /// Project `point` onto `polyline`, returning how far along the line the
    /// closest point is and how far off the line the fix sits.
    ///
    /// Uses a local equirectangular projection per polyline edge — accurate to
    /// well under GPS noise at segment scale (a few km), and cheap enough to
    /// run every second while racing. Must stay in lockstep with the JS twin
    /// in `ghost-race/tools/geo.mjs`; the shared `crosscheck.json` fixture
    /// pins both to the same expected values.
    public static func project(
        _ point: Coordinate,
        onto polyline: [Coordinate],
        cumulative: [Double]? = nil
    ) -> PolylineProjection {
        precondition(polyline.count >= 2, "polyline needs at least 2 points")
        let cums = cumulative ?? cumulativeDistances(polyline)

        var best = PolylineProjection(distanceAlongM: 0, offsetM: .greatestFiniteMagnitude)
        for i in 0..<(polyline.count - 1) {
            let a = polyline[i]
            let b = polyline[i + 1]
            let meanLat = (a.lat + b.lat) / 2 * .pi / 180
            let mPerDegLat = .pi / 180 * earthRadiusM
            let mPerDegLon = mPerDegLat * cos(meanLat)

            // Local planar coordinates in meters, origin at `a`.
            let bx = (b.lon - a.lon) * mPerDegLon
            let by = (b.lat - a.lat) * mPerDegLat
            let px = (point.lon - a.lon) * mPerDegLon
            let py = (point.lat - a.lat) * mPerDegLat

            let lenSq = bx * bx + by * by
            let t = lenSq == 0 ? 0 : max(0, min(1, (px * bx + py * by) / lenSq))
            let cx = t * bx
            let cy = t * by
            let offset = ((px - cx) * (px - cx) + (py - cy) * (py - cy)).squareRoot()
            if offset < best.offsetM {
                let edgeLen = (cums[i + 1] - cums[i])
                best = PolylineProjection(distanceAlongM: cums[i] + t * edgeLen, offsetM: offset)
            }
        }
        return best
    }
}
